src/amp-react-layout.js

```diff
@@ -0,0 +1,259 @@
+      inputLayout == Layout.FIXED_HEIGHT) &&
+    (!inputWidth || !inputHeight) &&
+    hasNaturalDimensions(tagName)
+  ) {
+    // Default width and height: handle elements that do not specify a
+    // width/height and are defined to have natural browser dimensions.
+    const dimensions = getNaturalDimensions(tagName);
+    width =
+      inputWidth || inputLayout == Layout.FIXED_HEIGHT
+        ? inputWidth
+        : dimensions.width;
+    height = inputHeight || dimensions.height;
+  } else {
+    width = inputWidth;
+    height = inputHeight;
+  }
+
+  // Calculate effective layout.
+  if (inputLayout) {
+    layout = inputLayout;
+  } else if (!width && !height) {
+    layout = Layout.CONTAINER;
+  } else if (height == 'fluid') {
+    layout = Layout.FLUID;
+  } else if (height && (!width || width == 'auto')) {
+    layout = Layout.FIXED_HEIGHT;
+  } else {
+    layout = Layout.FIXED;
+  }
+
+  // Apply UI.
+  const layoutStyle = {
+    'display': 'block',
+  };
+  const childStyle = {};
+  const layoutProps = {
+    'i-amphtml-layout': layout,
+    style: layoutStyle,
+  };
+  const layoutChildren = [];
+  if (isLayoutSizeDefined(layout)) {
+    layoutStyle['overflow'] = 'hidden';
+    layoutStyle['position'] = 'relative';
+    childStyle['position'] = 'absolute';
+    childStyle['inset'] = 0;
+  }
+  if (layout == Layout.NODISPLAY) {
+    // CSS defines layout=nodisplay automatically with `display:none`. Thus
+    // no additional styling is needed.
+    layoutProps['hidden'] = true;
+    // TODO(jridgewell): Temporary hack while SSR still adds an inline
+    // `display: none`
+    layoutStyle['display'] = '';
+  } else if (layout == Layout.FIXED) {
+    layoutStyle['width'] = width;
+    layoutStyle['height'] = height;
+  } else if (layout == Layout.FIXED_HEIGHT) {
+    layoutStyle['height'] = height;
+  } else if (layout == Layout.RESPONSIVE) {
+    layoutChildren.push(preact.createElement(
+      'i-amphtml-sizer',
+      {
+        style: {
+          'display': 'block',
+          'paddingTop':
+            (getLengthNumeral(height) / getLengthNumeral(width)) * 100 + '%',
+        },
+      }
+    ));
+  } else if (layout == Layout.FILL) {
+    // Do nothing.
+  } else if (layout == Layout.CONTAINER) {
+    // Do nothing. Elements themselves will check whether the supplied
+    // layout value is acceptable. In particular container is only OK
+    // sometimes.
+  } else if (layout == Layout.FLEX_ITEM) {
+    // Set height and width to a flex item if they exist.
+    // The size set to a flex item could be overridden by `display: flex` later.
+    if (width) {
+      layoutStyle['width'] = width;
+    }
+    if (height) {
+      layoutStyle['height'] = height;
+    }
+  } else {
+    // TODO: partial implementation.
+  }
+
+  // TODO: loading indicator.
+
+  return preact.createElement(
+    tagName || 'amp-element',
+    layoutProps,
+    layoutChildren.concat(
+      preact.createElement(
+        type,
```

Comment:
What is type?

Reply:
It's the type of the wrapper component. E.g. in 

```
<AmpWithLayout
  // Layout props
  layout="responsive"
  // Carousel's props, incl type
  type={AmpCarousel}
  currentSlide={2}
  >
...
```

This is a variation on the React's [higher order components](https://reactjs.org/docs/higher-order-components.html). But this is a JSX-compatible approach. I saw this approach recommended [here](https://stackoverflow.com/questions/49832457/how-to-add-additional-props-to-a-react-element-passed-in-as-a-prop). The higher order components themselves are used like this:

```
const AmpCarouselWithLayout = withLayout(AmpCarousel, ...);
```

We can do this as well, but I don't really like the idea of suffixing all components like this - but we can certainly explore that as well.

Naturally, I'd wish we could do something like this:

```
<AmpLayout layout="responsive" width={300} height={80}>
  <AmpCarousel></AmpCarousel>
</AmpLayout>
```

This would feel more like how the flexbox works. However, there's no "good" way to overwrite the `AmpCarousel`'s `style` prop in this way. It's possible to do via cloning, but I'd really rather not.

So the `type` option seems like a good compromise. From what I can tell it's fairly idiomatic, but it does feel a bit off.